README.md

```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
   - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
 - `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
 
 # Detailed design
 [design]: #detailed-design
 
-All attributes at the root of nixpkgs (`pkgs.<name>`) whose definition is of the form
+For all attributes at the root of nixpkgs `pkgs.<name>` which:
+1. Are defined in `pkgs/top-level/all-packages.nix`
+  (necessary so that the overlay containing the automatically discovered packages can be ordered directly before the `all-packages.nix` overlay without changing any behavior)
+2. Are defined to be equal to `pkgs.callPackage <path> { }` such that all transitively referenced paths from the default Nix file of `<path>`:
+  1. Are under the directory of `<path>`
+    (necessary so that moving these files to a new directory doesn't break references in this package)
+  2. Are not referenced from any paths outside of these transitive references
+    (necessary so that moving these files to a new directory doesn't break references in other packages)
+3. Evaluate to a derivation
+  (necessary because using `pkg-fun.nix` for a non-package would be counter-intuitive)
 
-```nix
-{
-  <name> = pkgs.callPackage ../some/dir { };
-}
-```
+These will be become eligible to be transformed as follows:
+- Move the default Nix file from `<path>` to `pkgs/unit/<4-prefix name>/<name>/pkg-fun.nix` ([TODO: Justify why `unit`](https://github.com/nixpkgs-architecture/simple-package-paths/issues/16))
+  - Where `<4-prefix name>` is the 4-letter prefix of `<name>`.
+    If `<name>` has less than 4 characters, append `-` to the `<name>` until it's 4 characters long and use that as `<4-prefix name>`
+- Additionally also move all paths transitively referenced by the default Nix file to `pkgs/unit/<4-prefix name>/<name>/???` [TODO](https://github.com/nixpkgs-architecture/simple-package-paths/issues/19)
+- Remove the definition of that attribute in `pkgs/top-level/all-packages.nix`
+- For each moved path, create a compatibility layer from the old to the new path, potentially using a symlink. See [compatibility] for more details
 
-will be become eligible to be transformed as follows:
-- Move `pkgs/some/dir` to `pkgs/auto/<name>`
-- Fix any references to that folder (e.g. in update scripts) 
-- Remove the original definition
+These attributes will newly be added to `pkgs` by automatically calling `pkgs.callPackage pkgs/unit/<4-prefix name>/<name>/pkg-fun.nix { }` on all entries in `pkgs/unit`. In order to make this more efficient, `builtins.readDir` should be optimized as described [here](https://github.com/NixOS/nix/issues/7314).
 
-These attributes will newly be defined by listing all directories in `pkgs/auto` using `builtins.readDir` and calling `pkgs.callPackage` on all of them, which then gets added to the `pkgs` scope.
+## Compatibility layer
+[compatibility]: #compatibility-layer
```

Comment:
This may be more trouble than it's worth.

Reply:
I don't think it's much trouble, it doesn't take a lot of effort to implement

Reply:
Any trouble at all is too much trouble of the effect isn't positive. By doing this we set a precedent that any file can be a public interface, while they're not supposed to.

Reply:
The file layout was not a public interface before. We just move packages and paths around with no considerations of breakages.

Reply:
I sort of agree with the sentiment of cost/benefit questions.

I think the main benefit is for people with am established muscle memory to quickly be redirected for the first couple of runs unil the muscle memory has adapted.

A one-line text file without any functionality other than informing the new location could just do.

No strong opinion, but maybe a way to scope this down a little.

Reply:
We discussed this [in yesterdays NAT meeting](https://discourse.nixos.org/t/2022-11-28-nixpkgs-architecture-team-meeting-19/23572#simple-package-paths-4). We concluded that we don't want a compatibility layer at all, because:
- The filesystem generally is not an API
  - While there's `pkgs.path`, it doesn't really make sense with overlays and config
  - There's `modulesPath` but that's specific to NixOS modules
  - There's `meta.position`, but only for debugging/better errors
- Especially with [the requirement for files to not be changed, only moved](https://github.com/nixpkgs-architecture/simple-package-paths/pull/20/files#r1035249558), not having a compatibility symlink works really nicely with PR's, as we believe git can handle rename merges properly

Reply:
Now removed the compatibility layer section, I added a mention to it in the alternatives section instead